### PR TITLE
Ball on net Evaluation functions (Plus relevant geom functions)

### LINKED ISSUE #298: Implement ball_on_net Evaluation Function
### Description of the task

Implement the ball_on_net evaluation function.

### Outline the task in more detail

We would like to implement the ball_on_net evaluation function in our new codebase. It is currently defined in the old codebase at `ai/hl/stp/evaluation/ball_threat.cpp`. With respect to the interface of this function (ie. what it takes and received), there should be similar types available in the new codebase. That said, it is possible that this function takes a type that is larger then it needs (ex. taking a `World` when it just needs a `Team` and a `ball`). Basically, try to use the smallest type you possible can; this will make it easier to test, review, develop, and work with in the future. 

**NOTE:** The current name for this function is poor, something like `ball_vel_intersects_friendly_net` would be a better alternative. 

At the same time, we should implement `ball_vel_intersects_enemy_net` since the math should be almost exactly the same.

### Acceptance criteria

- [ ] Function ball_on_net is implemented
- [ ] Function ball_on_net is unit tested

### PR DESCRIPTION
**Pull request by EvanMorcom**
Added detect_threat.cpp/h to evaluation functions. These files will be home to threat detection utilities. 
Implemented BallThreat class which contains friendly/enemy goal ball intersection functions. These functions can be used to detect if a ball must be blocked from entering our net, and if a ball is on a path that will end up in the enemy goal.
 
Implemented velocity line intersection function that will determine of the ball velocity will intersect a given line within a boundary rectangle. This can be used to determine if a ball will intersect friendly/enemy robots defined by a line, or if the ball will pass through a gap in robots. 

Unit tests were created for all newly introduced functions testing positive, negative, and edge cases.

Issues resolved #298 

### Review Checklist
- [X] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [X] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [X] **Remove all commented out code**
- [X] **Remove extra print statements**: for example, those just used for testing
- [X] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [X] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/geom/util.h [hunk 1]
```diff
@@ -338,6 +338,41 @@ std::vector<Point> lineIntersection(const Segment &a, const Segment &b);
...
 
+/**
+ * Calculates the intersection point (if it exists) of an moving element with a line
+ * defined by 2 Points
+ *
+ * @param element_velocity: X/Y velocity of the element, defined in meters per second
+ * @param element_position: X/Y position of the element, defined in meters
+ * @param line_start: X/Y point where the line begins, defined in meters
+ * @param line_end: X/Y point where the line ends, defined in meters
+ * @param boundry_area: Defines the rectangular geometry that limits the scope of
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.h]: Would be good to include more detail about what limiting the scope means here. Does the point of intersection need to be within the rectangle? Outside the rectangle? Does the `element_position` need to be inside? outside? etc.

File: src/thunderbots/software/geom/util.cpp [hunk 2]
```diff
@@ -752,6 +752,59 @@ std::optional<Point> lineIntersection(const Vector &a, const Vector &b, const Ve
...
+                                                const Rectangle boundry_area)
+{
+    // If the element is not within the boundries, return null
+    if (!boundry_area.containsPoint(element_position))
+    {
+        return std::nullopt;
+    }
+
+    // Calculate the closest edge of the boundry based on element speed and position
+    const double seconds_to_x_field_edge = std::max(
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: These variables should have "field" in them. The rectangle may not be referring to the field and this function has no concept of the field.

File: src/thunderbots/software/geom/util.cpp [hunk 3]
```diff
@@ -752,6 +752,59 @@ std::optional<Point> lineIntersection(const Vector &a, const Vector &b, const Ve
...
+                                                const Point line_end,
+                                                const Rectangle boundry_area)
+{
+    // If the element is not within the boundries, return null
+    if (!boundry_area.containsPoint(element_position))
+    {
+        return std::nullopt;
+    }
+
+    // Calculate the closest edge of the boundry based on element speed and position
```
Comment (MathewMacDougall) [inline src/thunderbots/software/geom/util.cpp]: Why does this need to be based on time? Isn't this just the intersection of the line formed by the velocity and the rectangle? Seems like we don't need the concept of units/time here (it's up to the caller to "know" what it's doing).

(same question for the `seconds_to_closest_edge` stuff below)

File: src/thunderbots/software/test/evaluation/detect_threat.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    EXPECT_EQ(std::nullopt, intersection);
+}
+
+
+int main(int argc, char **argv)
+{
+    std::cout << argv[0] << std::endl;
+    testing::InitGoogleTest(&argc, argv);
+    return RUN_ALL_TESTS();
+}
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/evaluation/detect_threat.cpp]: nit: missing newline at end of file

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Couple comments.

Also, all functions should be named `camelCaseWithLeadingLowercase` according to our convention.

File: src/thunderbots/software/test/evaluation/detect_threat.cpp (hunk 4 above)
Comment (EvanMorcom) [reply src/thunderbots/software/test/evaluation/detect_threat.cpp]: Added newline

File: src/thunderbots/software/geom/util.h (hunk 1 above)
Comment (EvanMorcom) [reply src/thunderbots/software/geom/util.h]: Good point, I elaborated on the scope.

File: src/thunderbots/software/geom/util.cpp (hunk 2 above)
Comment (EvanMorcom) [reply src/thunderbots/software/geom/util.cpp]: Good catch, I removed all references to time and field.

File: src/thunderbots/software/geom/util.cpp (hunk 3 above)
Comment (EvanMorcom) [reply src/thunderbots/software/geom/util.cpp]: I could definitely use line intersection functions here, but this seemed to me like the simplest way to find where the velocity will intersect the lines of the boundary rectangle. There exists a line-intersect-rectangle function but it considers all intersections as if the velocity line is infinite.

I've changed references to 'seconds' to 'proximity' as a catch-all term.

File: src/thunderbots/software/geom/util.h [hunk 5]
```diff
@@ -338,6 +338,42 @@ std::vector<Point> lineIntersection(const Segment &a, const Segment &b);
...
+ * calculating an intersection. Only intersections within boundry_area will be considered,
+ * and the element position must exist within the boundry.
+ */
+std::optional<Point> velocityLineIntersection(const Vector element_velocity,
+                                              const Vector element_position,
+                                              const Point line_start,
+                                              const Point line_end,
+                                              const Rectangle boundry_area);
```
Comment (garethellis0) [inline src/thunderbots/software/geom/util.h]: A few points here:
- I think you're looking for the concept of a `Ray` (ie. a vector with a start point)
- We should probably take in a `Segment` here, rather then two points
- Why do we have the boundary area?
- What about line segments that overlap the ray?

I took a quick stab at a revised function signature, let me know what you think:
```
/**
 * Finds the intersection of a given Ray and Segment, if there is one
 *
 * @param ray
 * @param segment
 *
 * @return
 *      - a tuple of (std::nullopt, std::nullopt) if there is no intersection point
 *      - a tuple of (point, std::nullopt) if there is one intersection point
 *      - a tuple of (point, point), if the ray and segment are parallel, where the two
 *        points are the end-points of the line segment made up by the overlap of the
 *        ray and line segment
 */
std::tuple<std::optional<Point>, std::optional<Point>> raySegmentIntersect(const Ray& ray, const Segment& segment);
```

File: src/thunderbots/software/geom/util.cpp [hunk 6]
```diff
@@ -752,6 +752,59 @@ std::optional<Point> lineIntersection(const Vector &a, const Vector &b, const Ve
...
+
+    // Grab the boundry edge that is intersected first
+    const double proximity_to_closest_edge =
+        std::min(proximity_to_x_boundry, proximity_to_y_boundry);
+
+    const Point boundry_edge_intersection =
+        proximity_to_closest_edge * element_velocity + element_position;
+
+    return finiteLengthLineIntersection(line_start, line_end, element_position,
+                                        boundry_edge_intersection);
```
Comment (garethellis0) [inline src/thunderbots/software/geom/util.cpp]: I think we could shrink the changes here and reduce the amount of code we have to add if we make use of some of the functions we already have here. I took a quick stab at an implementation that fits with the revised function signature I suggested, let me know what you think.
```
std::tuple<std::optional<Point>, std::optional<Point>> raySegmentIntersect(const Ray& ray, const Segment& segment){
    auto isect = lineIntersection(ray.getRayStart(), ray.getDirection(),
                                  segment.getSegStart(), segment.getEnd());

    // Check that there was an intersection and it is within the ray and segment
    if (isect.has_value() && contains(ray, isect.value()) && contains(segment, isect.value())) {
        // Check if the ray and segment are parallel
        if(collinear(ray.getRayStart(), ray.getDirection(), segment.getSegStart())){
            // The start point of the overlapping segment between the ray and segment
            Point overlapping_seg_start = closestPointOnSeg(ray.getRayStart(), segment);

            // TODO: Calc the second point for the overlapping segment. This will probably
            //       need some logic to account for different overlaps (segment lies
            //       entirely within ray, both partially overlap, etc)

            return {overlapping_seg_start, overlapping_seg_end};
        }

        // They are not parallel, so just return the intersection
        return {isect, std::nullopt};
    }

    return {std::nullopt, std::nullopt};
}

```

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/ai/hl/stp/evaluation/detect_threat.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
+/**
+ * This file contains independent Evaluation functions to evaluate threats to the friendly
+ * team on the playing field
+ */
+
+#ifndef THUNDERBOTS_ALL_DETECT_THREAT_H
+#define THUNDERBOTS_ALL_DETECT_THREAT_H
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/detect_threat.h]: Correct. Use `#pragma once`

File: src/thunderbots/software/geom/util.cpp (hunk 3 above)
Comment (EvanMorcom) [reply src/thunderbots/software/geom/util.cpp]: Changed functions (as suggested by Gareth), no longer applicable.

File: src/thunderbots/software/test/evaluation/detect_threat.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ */
+
+#include "ai/hl/stp/evaluation/detect_threat.h"
+
+#include <gtest/gtest.h>
+
+#include "ai/world/ball.h"
+#include "test/test_util/test_util.h"
+
+TEST(evaluation_detect_threat_test, ball_threat_ball_vel_to_friendly_net)
```
[the remark this answers is not in the document]
Comment (EvanMorcom) [reply src/thunderbots/software/test/evaluation/detect_threat.cpp]: Added short comments to new unit tests.

File: src/thunderbots/software/geom/util.h (hunk 5 above)
Comment (EvanMorcom) [reply src/thunderbots/software/geom/util.h]: Yeah this is a good idea. I implemented this function now.

File: src/thunderbots/software/geom/util.cpp (hunk 6 above)
Comment (EvanMorcom) [reply src/thunderbots/software/geom/util.cpp]: Implemented function.

**Comment by garethellis0** [comment]
@EvanMorcom - just as a note for the future, could you please leave comments unresolved and just leave a quick comment reply instead, indicating what you have/haven't changed? Just makes things a lot easier for me when I'm coming back and re-reviewing.

File: src/thunderbots/software/test/evaluation/detect_threat.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+#include "ai/hl/stp/evaluation/detect_threat.h"
+
+#include <gtest/gtest.h>
+
+#include "ai/world/ball.h"
+#include "test/test_util/test_util.h"
+
+// Test were the ball will intersect the friendly net
```
Comment (garethellis0) [inline src/thunderbots/software/test/evaluation/detect_threat.cpp]:
``` suggestion 
// Test where the ball will intersect the friendly net
```

File: src/thunderbots/software/geom/util.cpp [hunk 10]
```diff
@@ -752,6 +756,39 @@ std::optional<Point> lineIntersection(const Vector &a, const Vector &b, const Ve
...
+    {
+        return std::make_pair(intersection, std::nullopt);
+    }
+    // The ray and segment are parallel, and collinear
+    else if (!intersection.has_value() &&
+             collinear(ray.getRayStart(), segment.getSegStart(), segment.getEnd()))
+    {
+        // Since we know the ray and segment are overlapping, return the ray start
+        // position, and the end of the segment that is in the direction of the ray
+        ray.getDirection().norm() == (segment.getEnd() - ray.getRayStart()).norm()
```
Comment (garethellis0) [inline src/thunderbots/software/geom/util.cpp]: I might be wrong here, but this seems like it could have some issues with double precision such that this never returns true? Could we add a few more tests to check for the different overlapping cases to make sure this isn't an issue?

File: src/thunderbots/software/geom/util.cpp [hunk 11]
```diff
@@ -204,13 +203,18 @@ bool intersects(const Circle &first, const Circle &second)
...
+    // If the infinitely long vectors defined by ray and segment intersect, check that the
+    // intersection is within their definitions
     if (isect.has_value())
     {
         return contains(first, isect.value()) && contains(second, isect.value());
     }
-    return collinear(first.getRayStart(), first.getDirection(), second.getSegStart());
+    // If there is no intersection, the ray and segment may be parallel, check if they are
+    // overlapped
+    return contains(second, first.getRayStart());
```
Comment (garethellis0) [inline src/thunderbots/software/geom/util.cpp:217]: Why did we make changes here? If it was because this function was doing something incorrect, could we add a unit test for the behavior that was incorrect to ensure we don't regress? (In general we should just try to make sure that changes to functions are accompanied by corresponding test(s) or test(s) changes)

Same goes for the changes to `vectorRectIntersect` below.

**Comment by garethellis0** [comment]
This is starting to look really good! Just a few comments on the testing the changes we made to the geom functions and a possible floating point precision issue.

**Review by MathewMacDougall — COMMENTED** [review]
This is looking much better! I don't have anything to add on top of Gareth's comments.

File: src/thunderbots/software/geom/util.cpp (hunk 11 above)
Comment (EvanMorcom) [reply src/thunderbots/software/geom/util.cpp:217]: The function was improperly using Rays. It treats the ray direction as a point along the ray's path, which is incorrect. A point along the ray's path will be ray.start() + n*ray.direction()

Comment (EvanMorcom) [reply src/thunderbots/software/geom/util.cpp:217]: Added unit tests that checks the case that ray.getDirection intersects, but not (ray.getStartPosition() + ray.getDirection())

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: - just need to fix formatting as well

**Review by MathewMacDougall — DISMISSED** [review]
LGTM pending formatting :+1:

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1: Nice work. We can use this for the goalie soon :wink:
